src/deviation-check/deviation-check.ts

```diff
@@ -25,46 +29,39 @@ export const calculateMedian = (arr: bigint[]) => {
   return arr.length % 2 === 0 ? (nums[mid - 1]! + nums[mid]!) / 2n : nums[mid]!;
 };
 
-export const isDeviationThresholdExceeded = (onChainValue: bigint, deviationThreshold: bigint, apiValue: bigint) => {
-  const updateInPercentage = calculateDeviationPercentage(onChainValue, apiValue);
+export const isDeviationThresholdExceeded = (
+  onChainValue: bigint,
+  deviationThreshold: bigint,
+  apiValue: bigint,
+  deviationReference: bigint
+) => {
+  const updateInPercentage = calculateDeviationPercentage(onChainValue, apiValue, deviationReference);
 
-  return updateInPercentage > deviationThreshold;
+  return updateInPercentage >= deviationThreshold;
 };
 
-/**
- * Returns true when the on chain data timestamp is newer than the heartbeat interval.
- */
-export const isOnChainDataFresh = (timestamp: bigint, heartbeatInterval: bigint) =>
-  BigInt(timestamp) > BigInt(Math.floor(Date.now() / 1000)) - heartbeatInterval;
+export const isHeartbeatUpdatable = (timestamp: bigint, heartbeatInterval: bigint) =>
+  BigInt(timestamp) + heartbeatInterval <= BigInt(Math.floor(Date.now() / 1000));
 
+// Mirroring the logic of https://github.com/api3dao/airnode-protocol-v1/blob/65a77cdc23dc5434e143357a506327b9f0ccb7ef/contracts/api3-server-v1/extensions/BeaconSetUpdatesWithPsp.sol#L111
 export const isDataFeedUpdatable = (
   onChainValue: bigint,
   onChainTimestamp: bigint,
   offChainValue: bigint,
   offChainTimestamp: bigint,
   heartbeatInterval: bigint,
-  deviationThreshold: bigint
+  deviationThreshold: bigint,
+  deviationReference: bigint
 ): boolean => {
-  // Check that fulfillment data is newer than on chain data. Update transaction with stale data would revert on chain,
-  // draining the sponsor wallet. See:
-  // https://github.com/api3dao/airnode-protocol-v1/blob/dev/contracts/dapis/DataFeedServer.sol#L121
-  if (offChainTimestamp <= onChainTimestamp) {
```

Comment:
We surely need this check. Because airseeker-v2 utilizes delayed endpoint, there can be some scenario like:
1. One updates data feed with OEV which is zero-delayed data
2. Airseeker has 15 secs delayed data which can exceed threshold
3. Airseeker tries to update feed but couldn't be successful because [this](https://github.com/api3dao/contracts/blob/phase21/contracts/api3-server-v1/DataFeedServer.sol#L120) and it results in unnecessarily spent from sponsor wallet.

Reply:
Before doing step 3) we check which beacons need to be updated [here](https://github.com/api3dao/airseeker-v2/blob/0ac8e98a4f90fb15dbc8b1f5fb20dabfc1ea1e85/src/update-feeds-loops/get-updatable-feeds.ts#L54) and  would use the on-chain value (provided we fetched it). In case we don't have the latest on-chain data, then Airseeker may attempt an update which would revert - this might fail [here](https://github.com/api3dao/airseeker-v2/blob/0ac8e98a4f90fb15dbc8b1f5fb20dabfc1ea1e85/src/update-feeds-loops/submit-transactions.ts#L168), but if there is a fallback gas limit provided, the transaction may indeed by submitted.

OEV updates are proxy specific, so that doesn't matter - but we may be running backup Airseekers so we should assume the feed can be updated from other sources, so your point is still valid.

But your comment made me think of another case, which I missed previously:
1. Say there is beacon set with 3 beacons.
2. Say Airseeker updates all beacons and the beacon set.
3. Now Signed API goes down, so there are no new data. This persists for a a couple of minutes.
4. None of the beacons are updatable, but the beacon set is because of the heartbeat.
5. We call [this](https://github.com/api3dao/airnode-protocol-v1/blob/65a77cdc23dc5434e143357a506327b9f0ccb7ef/contracts/api3-server-v1/DataFeedServer.sol#L46) without updating any beacon which will revert - although we could have prevent this.

Note that this is only for the beacon set, but yeah - we need the check for equality (but not for being smaller).

Reply:
There is a similar edge case when the data feed is a beacon. What ends up happening we submit a multi call transaction with empty calldata.

Reply:
Take a look at https://github.com/api3dao/airseeker-v2/pull/233/commits/6bd46b286769d89b595e7d22aa1c8344c04522a7

I think that mimics the contract logic. Note, that the timestamp check (as it was before) is more strict compared to the contract, because the contract allows the beacon set also if value change (and timestamp stays the same).

Reply:
I think that the most accurate implementation has been reached. I've thought about it quite and couldn't find anything to worry about 💪🏼.